File: engine-core/src/threats.cpp
#include "gomoku/threats.hpp"

#include <cstdint>

namespace gomoku {

// =========================================================================
// Internal helpers
// =========================================================================

// Find all moves that create an immediate win (five in a row) for `player`.
// This is equivalent to JS findWinningMoves: try each empty cell, check if
// placing there completes a line.
static std::vector<int> find_winning_moves(const GomokuBoard& board, int8_t player) {
    const int N = board.board_size();
    const int total = N * N;
    const int win_len = board.win_len();
    std::vector<int> moves;

    static constexpr int dr[4] = {0, 1, 1, 1};
    static constexpr int dc[4] = {1, 0, 1, -1};

    for (int pos = 0; pos < total; pos++) {
        if (board.cell_at(pos) != EMPTY) continue;
        int r0 = pos / N, c0 = pos % N;
        bool wins = false;
        for (int d = 0; d < 4 && !wins; d++) {
            int count = 1;  // the stone we'd place
            for (int sign = -1; sign <= 1; sign += 2) {
                for (int step = 1; step < win_len; step++) {
                    int rr = r0 + sign * step * dr[d];
                    int cc = c0 + sign * step * dc[d];
                    if (rr < 0 || rr >= N || cc < 0 || cc >= N) break;
                    if (board.cell_at(rr * N + cc) != player) break;
                    count++;
                }
            }
            if (count >= win_len) wins = true;
        }
        if (wins) moves.push_back(pos);
    }
    return moves;
}

// Check if a move is winning for the player whose stone is at that cell.
// Used after make_move to verify a win.
static bool is_winning_at(const GomokuBoard& board, int pos, int8_t player) {
    const int N = board.board_size();
    const int win_len = board.win_len();
    int r0 = pos / N, c0 = pos % N;

    static constexpr int dr[4] = {0, 1, 1, 1};
    static constexpr int dc[4] = {1, 0, 1, -1};

    for (int d = 0; d < 4; d++) {
        int count = 1;
        for (int sign = -1; sign <= 1; sign += 2) {
            for (int step = 1; step < win_len; step++) {
                int rr = r0 + sign * step * dr[d];
                int cc = c0 + sign * step * dc[d];
                if (rr < 0 || rr >= N || cc < 0 || cc >= N) break;
                if (board.cell_at(rr * N + cc) != player) break;
                count++;
            }
        }
        if (count >= win_len) return true;
    }
    return false;
}

// =========================================================================
// find_four_creating_moves_optimized
// =========================================================================

std::vector<int> find_four_creating_moves_optimized(GomokuBoard& board, int8_t player) {
    const int N = board.board_size();
    const int win_len = board.win_len();
    const int total = N * N;
    std::vector<int> moves;
    std::vector<uint8_t> seen(total, 0);

    static constexpr int dr[4] = {0, 1, 1, 1};
    static constexpr int dc[4] = {1, 0, 1, -1};

    for (int pos = 0; pos < total; pos++) {
        if (board.cell_at(pos) != EMPTY || seen[pos]) continue;

        // Temporarily place the stone using make_move.
        // IMPORTANT: make_move uses side_to_move_. We need to place for `player`.
        // Since the board alternates sides, we must work at cell level.
        // But we don't have direct cell write access. Instead, we'll do the
        // analysis using only cell_at reads and manual line scanning.

        // Approach: simulate placement by reading neighbors. If pos were `player`:
        int r = pos / N, c = pos % N;
        bool creates_four = false;

        for (int d = 0; d < 4 && !creates_four; d++) {
            int fwd = 0, bwd = 0;
            int rr, cc;

            // Forward
            rr = r + dr[d]; cc = c + dc[d];
            while (rr >= 0 && cc >= 0 && rr < N && cc < N &&
                   board.cell_at(rr * N + cc) == player) {
                fwd++; rr += dr[d]; cc += dc[d];
            }
            bool fwd_empty = (rr >= 0 && cc >= 0 && rr < N && cc < N &&
                              board.cell_at(rr * N + cc) == EMPTY);

            // Backward
            rr = r - dr[d]; cc = c - dc[d];
            while (rr >= 0 && cc >= 0 && rr < N && cc < N &&
                   board.cell_at(rr * N + cc) == player) {
                bwd++; rr -= dr[d]; cc -= dc[d];
            }
            bool bwd_empty = (rr >= 0 && cc >= 0 && rr < N && cc < N &&
                              board.cell_at(rr * N + cc) == EMPTY);

            int len = 1 + fwd + bwd;

            // Creates five directly
            if (len >= win_len) {
                creates_four = true;
                break;
            }

            // Creates a four (winLen-1 in a row with at least one open end)
            if (len == win_len - 1 && (fwd_empty || bwd_empty)) {
                creates_four = true;
                break;
            }

            // Check for gapped fours: windows of size winLen that include pos
            for (int offset = 0; offset < win_len && !creates_four; offset++) {
                int start_r = r - offset * dr[d];
                int start_c = c - offset * dc[d];
                int p_count = 0, e_count = 0;
                bool valid = true;

                for (int k = 0; k < win_len; k++) {
                    int kr = start_r + k * dr[d];
                    int kc = start_c + k * dc[d];
                    if (kr < 0 || kc < 0 || kr >= N || kc >= N) {
                        valid = false; break;
                    }
                    int idx = kr * N + kc;
                    int8_t v;
                    if (idx == pos) {
                        v = player;  // simulated placement
                    } else {
                        v = board.cell_at(idx);
                    }
                    if (v == player) p_count++;
                    else if (v == EMPTY) e_count++;
                    else { valid = false; break; }
                }

                if (valid && p_count == win_len - 1 && e_count == 1) {
                    creates_four = true;
                }
            }
        }

        if (creates_four) {
            moves.push_back(pos);
            seen[pos] = 1;
        }
    }

    return moves;
}

// =========================================================================
// VCF search (recursive)
// =========================================================================

struct VcfState {
    int nodes_searched = 0;
    int max_nodes = 50000;
};

// Recursive VCF search. Returns the winning sequence or empty vector.
static std::vector<int> vcf_recurse(GomokuBoard& board, int8_t attacker,
                                     int depth, std::vector<int>& sequence,
                                     VcfState& state) {
    if (state.nodes_searched++ > state.max_nodes) return {};
    if (depth <= 0) return {};

    // Attacker's turn: find four-creating moves
    auto four_moves = find_four_creating_moves_optimized(board, attacker);

    for (int attack_move : four_moves) {
        // Make the attack move. Note: board.make_move uses side_to_move_.
        // We need to ensure side_to_move_ matches attacker.
        // The VCF alternates attacker -> defender -> attacker, and the
        // board side also alternates. As long as we start with correct side,
        // it stays in sync.
        const int8_t defender = opponent(attacker);
        board.make_move(attack_move, attacker);

        // Check if this move itself is a win (five in a row)
        if (is_winning_at(board, attack_move, attacker)) {
            board.undo_move();
            sequence.push_back(attack_move);
            return sequence;
        }

        // Find the threat(s) this creates
        auto threats = find_winning_moves(board, attacker);

        if (static_cast<int>(threats.size()) >= 2) {
            // Double threat = immediate win (opponent can block only one)
            board.undo_move();
            sequence.push_back(attack_move);
            return sequence;
        }

        if (threats.size() == 1) {
            // Single threat — opponent must block
            int block_move = threats[0];
            board.make_move(block_move, defender);

            // After block, check if game continues
            Cell w = board.winner();
            if (w == GomokuBoard::ONGOING) {  // game still in progress
                // Continue: attacker plays next four
                sequence.push_back(attack_move);
                sequence.push_back(block_move);
                auto result = vcf_recurse(board, attacker, depth - 2, sequence, state);
                if (!result.empty()) {
                    board.undo_move();  // undo block
                    board.undo_move();  // undo attack
                    return result;
                }
                // Remove the moves we added
                sequence.pop_back();
                sequence.pop_back();
            }

            board.undo_move();  // undo block
        }

        board.undo_move();  // undo attack
    }

    return {};  // no winning sequence found
}

// =========================================================================
// Public API
// =========================================================================

ThreatResult vcf_search(GomokuBoard& board, int8_t attacker,
                        int max_depth, int max_nodes) {
    ThreatResult result;

    // Quick check: if attacker already has a winning move, return it
    auto immediate_wins = find_winning_moves(board, attacker);
    if (!immediate_wins.empty()) {
        result.found = true;
        result.move = immediate_wins[0];
        result.sequence = {immediate_wins[0]};
        result.nodes_searched = 1;
        return result;
    }

    // With the new API, make_move takes an explicit player argument,
    // so no side_to_move check is needed.

    VcfState state;
    state.max_nodes = max_nodes;

    std::vector<int> sequence;
    auto seq = vcf_recurse(board, attacker, max_depth, sequence, state);
    result.nodes_searched = state.nodes_searched;

    if (!seq.empty()) {
        result.found = true;
        result.move = seq[0];
        result.sequence = std::move(seq);
    }

    return result;
}

int find_best_vcf_defense(GomokuBoard& board, int8_t defender, int8_t attacker,
                          const std::vector<int>& candidates, int vcf_max_depth) {
    for (int def_move : candidates) {
        if (board.cell_at(def_move) != EMPTY) continue;

        board.make_move(def_move, defender);

        // Check if attacker still has VCF with reduced depth
        auto attack_vcf = vcf_search(board, attacker, vcf_max_depth - 2, 50000);

        board.undo_move();

        if (!attack_vcf.found) {
            // This move breaks the VCF
            return def_move;
        }
    }
    return -1;  // no defense found
}

bool has_immediate_win(GomokuBoard& board, int8_t player) {
    auto wins = find_winning_moves(board, player);
    return !wins.empty();
}

int count_four_threats(GomokuBoard& board, int8_t player) {
    // Count empty cells where placing player's stone would create five
    auto wins = find_winning_moves(board, player);
    return static_cast<int>(wins.size());
}

} // namespace gomoku

File: server/src/engine/threats.mjs
// Threat Space Search: VCF (Victory by Continuous Fours)
// Finds forced winning sequences through continuous four-threats
// The defender has exactly one response to each four (must block)

import { findWinningMoves } from './patterns.mjs';
import { THREAT_CFG } from './config.mjs';

/**
 * VCF Search — find a forced win by playing only four-threats.
 *
 * A "four" is a move that creates an immediate win on the NEXT move.
 * The opponent MUST respond by blocking (exactly one blocking move per four).
 *
 * The search alternates:
 *   1. Attacker plays a four-creating move
 *   2. Defender is forced to block
 *   3. Repeat until attacker achieves FIVE or depth limit
 *
 * @param {GomokuBoard} board
 * @param {number} attacker - +1 or -1
 * @param {number} [maxDepth] - max depth in half-moves (attack + defense pairs)
 * @returns {{ move: number, sequence: number[] } | null} winning first move or null
 */
export function vcfSearch(board, attacker, maxDepth) {
  maxDepth = maxDepth ?? THREAT_CFG.vcfMaxDepth;
  let nodesSearched = 0;
  const maxNodes = THREAT_CFG.vcfMaxNodes;

  /**
   * Recursive VCF search.
   * @param {number} depth - remaining depth
   * @param {boolean} isAttacker - true = attacker's turn
   * @param {number[]} sequence - move sequence so far
   * @returns {number[]|null} winning sequence or null
   */
  function vcfRecurse(depth, isAttacker, sequence) {
    if (nodesSearched++ > maxNodes) return null;
    if (depth <= 0) return null;

    const cells = board.cells;

    if (isAttacker) {
      // Find moves that create an immediate winning threat (a "four")
      // A four-creating move is one where, after placement, there exists
      // at least one cell where placing another stone wins
      const fourMoves = findFourCreatingMovesOptimized(board, attacker);

      for (const attackMove of fourMoves) {
        board.makeMove(attackMove, attacker);

        // Check if this move itself is a win (five in a row)
        if (board.isWinningMove(attackMove, attacker)) {
          board.undoMove();
          return [...sequence, attackMove];
        }

        // Find the threat(s) this creates
        const threats = findWinningMoves(board, attacker);

        if (threats.length >= 2) {
          // Double threat = immediate win (opponent can block only one)
          board.undoMove();
          return [...sequence, attackMove];
        }

        if (threats.length === 1) {
          // Single threat — opponent must block at threats[0]
          const blockMove = threats[0];

          board.makeMove(blockMove, -attacker);

          // After block, check if game continues
          const winner = board.winner();
          if (winner === null) {
            // Continue: attacker plays next four
            const result = vcfRecurse(depth - 2, true, [...sequence, attackMove, blockMove]);
            if (result) {
              board.undoMove(); // undo block
              board.undoMove(); // undo attack
              return result;
            }
          }

          board.undoMove(); // undo block
        }

        board.undoMove(); // undo attack
      }

      return null;
    }

    return null; // Should not reach here
  }

  // Quick check: if attacker already has a winning move, return it
  const immediateWins = findWinningMoves(board, attacker);
  if (immediateWins.length >= 1) {
    return { move: immediateWins[0], sequence: [immediateWins[0]] };
  }

  const result = vcfRecurse(maxDepth, true, []);
  if (result && result.length > 0) {
    return { move: result[0], sequence: result };
  }

  return null;
}

/**
 * Optimized four-creating move finder.
 * A four-creating move is one where, after placing a stone:
 * - The move itself forms part of a line of (winLen-1) with at least one open end, OR
 * - The move creates a new winning threat on the next move
 *
 * We use a fast line-scan approach instead of brute-force.
 */
function findFourCreatingMovesOptimized(board, player) {
  const N = board.N;
  const winLen = board.winLen;
  const cells = board.cells;
  const moves = [];
  const seen = new Uint8Array(board.size);

  const DIRS = [[0, 1], [1, 0], [1, 1], [1, -1]];

  for (let pos = 0; pos < board.size; pos++) {
    if (cells[pos] !== 0 || seen[pos]) continue;

    // Temporarily place
    cells[pos] = player;

    // Check each direction: does this create a line of winLen-1 with open end?
    let createsFour = false;
    const r = (pos / N) | 0;
    const c = pos % N;

    for (const [dr, dc] of DIRS) {
      let fwd = 0, bwd = 0;
      let rr, cc;

      // Forward
      rr = r + dr; cc = c + dc;
      while (rr >= 0 && cc >= 0 && rr < N && cc < N && cells[rr * N + cc] === player) {
        fwd++; rr += dr; cc += dc;
      }
      const fwdEmpty = rr >= 0 && cc >= 0 && rr < N && cc < N && cells[rr * N + cc] === 0;

      // Backward
      rr = r - dr; cc = c - dc;
      while (rr >= 0 && cc >= 0 && rr < N && cc < N && cells[rr * N + cc] === player) {
        bwd++; rr -= dr; cc -= dc;
      }
      const bwdEmpty = rr >= 0 && cc >= 0 && rr < N && cc < N && cells[rr * N + cc] === 0;

      const len = 1 + fwd + bwd;

      // Creates five directly
      if (len >= winLen) {
        createsFour = true;
        break;
      }

      // Creates a four (winLen-1 in a row with at least one open end)
      if (len === winLen - 1 && (fwdEmpty || bwdEmpty)) {
        createsFour = true;
        break;
      }

      // Check for gapped fours: e.g., XX_XX where filling pos creates X_XXXX or XXXX_X
      // Check windows of size winLen that include pos
      for (let offset = 0; offset < winLen; offset++) {
        const startR = r - offset * dr;
        const startC = c - offset * dc;
        let pCount = 0, eCount = 0, valid = true;

        for (let k = 0; k < winLen; k++) {
          const kr = startR + k * dr;
          const kc = startC + k * dc;
          if (kr < 0 || kc < 0 || kr >= N || kc >= N) { valid = false; break; }
          const v = cells[kr * N + kc];
          if (v === player) pCount++;
          else if (v === 0) eCount++;
          else { valid = false; break; }
        }

        if (valid && pCount === winLen - 1 && eCount === 1) {
          createsFour = true;
          break;
        }
      }

      if (createsFour) break;
    }

    cells[pos] = 0; // restore

    if (createsFour) {
      moves.push(pos);
      seen[pos] = 1;
    }
  }

  return moves;
}

/**
 * Find the best defensive move against an opponent's VCF.
 *
 * Strategy: try each candidate move for the defender, and check if
 * the attacker's VCF still works. Return the move that breaks the VCF.
 *
 * @param {GomokuBoard} board
 * @param {number} defender - defending player
 * @param {number} attacker - attacking player
 * @param {number[]} candidates - candidate defensive moves
 * @returns {number} best defensive move, or -1 if none found
 */
export function findBestVCFDefense(board, defender, attacker, candidates) {
  // First, try all candidate moves
  for (const defMove of candidates) {
    if (board.cells[defMove] !== 0) continue;

    board.makeMove(defMove, defender);

    // Check if attacker still has VCF with reduced depth
    const attackVcf = vcfSearch(board, attacker, THREAT_CFG.vcfMaxDepth - 2);

    board.undoMove();

    if (!attackVcf) {
      // This move breaks the VCF!
      return defMove;
    }
  }

  return -1; // no defense found
}

/**
 * Quick check: does a player have any immediate winning move?
 */
export function hasImmediateWin(board, player) {
  return findWinningMoves(board, player).length > 0;
}

/**
 * Count how many separate four-threats exist for a player.
 * Useful for detecting double-threat (fork) situations.
 */
export function countFourThreats(board, player) {
  const cells = board.cells;
  let count = 0;

  for (let pos = 0; pos < board.size; pos++) {
    if (cells[pos] !== 0) continue;
    cells[pos] = player;
    if (board.isWinningMove(pos, player)) {
      count++;
    }
    cells[pos] = 0;
  }

  return count;
}
